Bugfix: invalid parsing of binlog event data

andreas-bok-sociomantic opened the pull request:
The fixed bytesize binlog header was incorrectly
defined as 27.

ben-palmer-sociomantic commented:
Should the comparison here https://github.com/andreas-bok-sociomantic/libdrizzle-redux/blob/feeb4f7b1405f05c99e604c61e47c3049a28b0b6/libdrizzle/binlog.cc#L337 also be adjusted?

andreas-bok-sociomantic commented:
> Should the comparison here https://github.com/andreas-bok-sociomantic/libdrizzle-redux/blob/feeb4f7b1405f05c99e604c61e47c3049a28b0b6/libdrizzle/binlog.cc#L337 also be adjusted?

Looked at the MySQL binlog documentation again. AFAICS no event will ever have a length <=27 for all version of the binlog, https://dev.mysql.com/doc/internals/en/binary-log-versions.html 

I tried adding an assert(false) in the first line of the section. Binlog unittest still succeeds. Thus quite sure the code will never be reached and can be safely removed.

ben-palmer-sociomantic commented:
> I tried adding an assert(false) in the first line of the section. Binlog unittest still succeeds. Thus quite sure the code will never be reached and can be safely removed.

Ok maybe just change it to an assert that the length is at least the size of the header or whatever else is the minimum required (19?).

Is this related to https://github.com/sociomantic-tsunami/libdrizzle-redux/issues/214 ?

andreas-bok-sociomantic commented:
> Ok maybe just change it to an assert that the length is at least the size of the header or whatever else is the minimum required (19?)

Yes, I would also prefer to keep the code and add an assert

> Is this related to #214 ?

No

andreas-bok-sociomantic committed:
Bugfix: Invalid parsing of binlog event packet data

Correctly set the length of the event data to be
size of the whole binlog event minus the size of
header.

The bug was caused by an extra 8 bytes
offset which corrupted the pointer to the
binlog event data

andreas-bok-sociomantic commented:
Looked at the documentation again. It is actually the case that event types **APPEND_BLOCK_EVENT, EXEC_LOAD_EVENT, DELETE_FILE_EVENT** have length of 27. The layout is 
```
Fixed Header                  19
Fixed data part                4
Variable data part             0
Checksum                       4

```
https://dev.mysql.com/doc/internals/en/load-data-infile-events.html

I would like to look further into this issue, and thus remove the change from this PR.

andreas-bok-sociomantic committed:
Use enum in ptr arithmetic when parsing binlog data

Use the existing enum drizzle_binlog_event_positions_t
which specifies the start index of the different parts
of the fixed header

ben-palmer-sociomantic commented:
> I would like to look further into this issue before adding an assert(false).

Cool I blocked it just to be clear.

andreas-bok-sociomantic commented:
> Cool I blocked it just to be clear.

Cool, the current changes though are good to go, so I will unblock.

ben-palmer-sociomantic commented:
LGTM.
